### PR TITLE
reduce transfers between host and gpus. 

### PR DESCRIPTION
**Pull request by gvessere**
The host is responsible for splitting the batch and each GPU gets only the data it will work on (as opposed to the entire batch).

This also probably has the benefit of saving some memory on the GPU side.

### CODE REVIEW

File: keras/utils/training_utils.py [hunk 1]
```diff
@@ -126,7 +126,7 @@ def get_slice(data, i, parts):
     # Place a copy of the model on each GPU,
     # each getting a slice of the inputs.
     for i in range(gpus):
-        with tf.device('/gpu:%d' % i):
+        with tf.device('/cpu:%d' % i):
```
Comment (smgt) [inline keras/utils/training_utils.py]: Should'nt it be `/cpu:0`? Have tensorflow implemented anything else?

Comment (fchollet) [reply keras/utils/training_utils.py]: Yes, it should

**Comment by gvessere** [comment]
ok, thx, that's fixed.

**Comment by fchollet** [comment]
I tested this, and I observe no measurable change with a range of models (Xception, ResNet50, VGG16) and a range of GPU count (2, 4, 8). We're consistently the same speed, actually 1ms slower per batch a lot of the time (which could be noise).

Under which conditions did you observe a speed up?

**Comment by gvessere** [comment]
I have 3D data with 128x128x128 float cubes and batch size of 128. Which amounts to about 1 Gbyte per batch. If you're testing with 2d data you might want to rescale your images to 1450x1450 and increase the stride on the first Conv to have reasonably sized feature maps.

**Comment by gvessere** [comment]
oh, and this is a 4 GPU setup.

**Comment by fchollet** [comment]
Can you post:

- a demo model (standalone script, data randomly generated with numpy)
- timing you're observing with and without this PR
- memory usage you're observing with and without this PR

I will try to repro with your script with various GPUs

**Comment by gvessere** [comment]
Upon further investigation, it seems the speed-up I observed was probably related to something else like maybe disk caching, I'm not too sure. I did dig a bit further into whether the mod was following my intuition, and it is not quite right. As François mentioned there is a slight drop in performance that is consistent with what I observe in the GPU profiler (see image of profiler on modded version). The original version shows a full batch transfer to all GPUs (sorry I lost that view), after which each GPU extracts what it needs (1/4 batch on my 4 GPU setup). My modded version transfers the whole batch to GPU0, then transfers 1/4 batch to GPU0 again. GPUs 1 to 3 get their batch data from GPU0 (see image). I will close this PR, as it doesn't improve things. Also to put things in context, one full batch (1Gbytes in my case) takes up to half a second to transfer to gpus, so the potential gain if we do this right would be about 150-200ms per batch.

image

**Pull request closed without merging**